### ISSUE TITLE
twisted.web.server.Request.view_setResponseCode doesn't accept message parameter

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#5525]
|[<img alt="tomprince's avatar" src="https://avatars.githubusercontent.com/u/283816?s=50" width="50" height="50">](https://github.com/tomprince)| @tomprince reported|
|-|-|
|Trac ID|trac#5525|
|Type|defect|
|Created|2012-03-12 04:01:25Z|
|Branch|https://github.com/twisted/twisted/tree/setresponsecode-distrib-5525|

Attachments:

* [distrib_server_response_code_message.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/f95/f95145d3b7a732187ccdf248fa5038d8490b92ed/a87d55d8cb698c6570ab8b8399646f8cda69e1aa.patch) (2132 bytes) - added by tom.prince on 2012-03-12 04:01:54Z - Add missing message parameter
* [distrib_server_response_code_message_2.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/f95/f95145d3b7a732187ccdf248fa5038d8490b92ed/affe3b165531bfc70f4d3e4c5a31a1f540239ed7.patch) (5392 bytes) - added by tom.prince on 2012-03-13 02:50:38Z - patch v2
* [distrib_server_response_code_message_3.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/f95/f95145d3b7a732187ccdf248fa5038d8490b92ed/2d10aad9edb2b25e17072ea8049c7cc081180656.patch) (6029 bytes) - added by tom.prince on 2012-03-15 14:45:28Z - Patch v3.

<details><summary>Searchable metadata</summary>

```
trac-id__5525 5525
type__defect defect
reporter__tom_prince tom.prince
priority__normal normal
milestone__ 
branch__branches_setresponsecode_distrib_5525 branches/setresponsecode-distrib-5525
branch_author__exarkun exarkun
status__closed closed
resolution__fixed fixed
component__web web
keywords__ 
time__1331524885000000 1331524885000000
changetime__1331927835000000 1331927835000000
version__None None
owner__exarkun exarkun
cc__jknight cc__jesstess
```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="jesstess's avatar" src="https://avatars.githubusercontent.com/u/188336?s=50" width="50" height="50">](https://github.com/jesstess)|@jesstess set owner to @jesstess|
|-|-|

**Comment by twisted-trac** [comment]
|<img alt="Fahrenheit's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Fahrenheit set owner to Fahrenheit<br>Fahrenheit set status to `assigned`|
|-|-|

**Comment by twisted-trac** [comment]
|<img alt="Fahrenheit's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_4"></a>|Fahrenheit set owner to @tomprince<br>Fahrenheit set status to `new`|
|-|-|

Thanks for the patch! Comments below:
  1. `twisted/web/server.py`
    1. Not a huge deal, but the function signature for `view_setResponseCode` should match that of `setResponseCode`, so maybe remove the whitespace around the '='.
  1. `twisted/web/test/test_distrib.py`
    1. Again, not a huge deal, but the second underscore in `test_setResponseCode_message` doesn't match the twisted guidelines; everything after "test_" should be camelcase, rather that separated by any underscores.
    1. Last issue, and this one matters, is that your test doesn't actually make sure that `view_setResponseCode` does the correct thing with the `message` argument; I was able to remove it from your implementation and the test still passed, so make sure that your test checks that the message gets used properly.

You should be good once you fix up the test and the style stuff would be nice too, while you're at it.  Thanks again!

**Comment by twisted-trac** [comment]
|[<img alt="tomprince's avatar" src="https://avatars.githubusercontent.com/u/283816?s=50" width="50" height="50">](https://github.com/tomprince)<a name="note_5"></a>|@tomprince commented|
|-|-|

I think this fixes all issues. The old _requestTest used getPage, which didn't have a way of looking
at the response. This changes the tests to use Agent.

**Comment by twisted-trac** [comment]
|[<img alt="tomprince's avatar" src="https://avatars.githubusercontent.com/u/283816?s=50" width="50" height="50">](https://github.com/tomprince)|@tomprince removed owner|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="jesstess's avatar" src="https://avatars.githubusercontent.com/u/188336?s=50" width="50" height="50">](https://github.com/jesstess)|@jesstess set owner to @jesstess|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="jesstess's avatar" src="https://avatars.githubusercontent.com/u/188336?s=50" width="50" height="50">](https://github.com/jesstess)<a name="note_8"></a>|@jesstess set owner to @tomprince|
|-|-|

Thanks for working on this, tom.prince. A few comments:

 * Since `view_setResponseCode` is in `twisted.web.server`, it seems like tests for this change would go in `twisted.web.test.test_web` with the other server tests. Can you explain why they should be in `test_distrib`, or move them to `test_web`? There's a `test_setResponseCode` in `test_http` you could crib from. It would also allow you to construct a more succinct unit test than one that is forcing you to make changes in `test_largeWrite` and `test_largeReturn`.
 * If your changes stay in `test_distrib`, there are a couple of typos:
  * The test no longer uses `getPage`:
```
        @param **kwargs: Extra keyword arguments to pass to L{getPage} when
```
  * consiting ==> consisting in:
```
        @return: A L{Deferred} which fires with a tuple consiting of a                                                                                                                                           
```
  * IResposne ==> IResponse in:
```
            body of the response and an L{IResposne} with the response itself.
```

Sorry for the back and forth, and thanks again for working on this!

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_9"></a>|@exarkun commented|
|-|-|

> Since view_setResponseCode is in twisted.web.server, it seems like tests for this change would go in twisted.web.test.test_web

The strange factoring of the implementation of distrib means that the test organization differs a bit from the usual approach.  I think `test_distrib` is the right place for these tests, because that's where there's testing infrastructure for actually exercising the distrib system.  That involves setting up a server and connecting a distrib client to it (which would be nice to avoid, but that's how all the existing distrib tests are written...).

**Comment by twisted-trac** [comment]
|[<img alt="tomprince's avatar" src="https://avatars.githubusercontent.com/u/283816?s=50" width="50" height="50">](https://github.com/tomprince)<a name="note_10"></a>|@tomprince removed owner|
|-|-|

The typos have been fixed. I also split _requestTest, so that creating the servers is separate from making the request, allowing the existing tests to remain unchanged.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_11"></a>|@exarkun commented|
|-|-|

(In [33876]) Branching to 'setresponsecode-distrib-5525'

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_12"></a>|@exarkun commented|
|-|-|

(In [33877]) Apply distrib_server_response_code_message_3.patch 

refs [#5525](https://github.com/twisted/twisted/issues/5525)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)|@exarkun set owner to @exarkun<br>@exarkun set status to `assigned`|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_14"></a>|@exarkun commented|
|-|-|

Thanks.  This is looking pretty great now.  Just a few minor points:

  1. Methods in a class should be separated from each other by two blank lines
  1. There are some tabs in the patch; there should never be tabs in a .py file.
  1. The change needs a [news fragment](https://twistedmatrix.com/trac/wiki/ReviewProcess#Newsfiles).  This is a nice bug fix for distrib servers, so it's worth mentioning in a .bugfix fragment.

These are pretty trivial points.  If I get a clean build on buildbot, I'll deal with these changes and merge.  Thanks again!

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_15"></a>|@exarkun commented|
|-|-|

[Build results](http://buildbot.twistedmatrix.com/boxes-supported?branch=/branches/setresponsecode-distrib-5525)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_16"></a>|@exarkun set status to `closed`|
|-|-|

(In [33885]) Mege setresponsecode-distrib-5525 

Author: tom.prince
Reviewer: Fahrenheit, jesstess, exarkun
Fixes: [#5525](https://github.com/twisted/twisted/issues/5525)

Add the `message` parameter to the PB view of `twisted.web.server.Request`.
This allows distributed servers (`twisted.web.distrib`) to set the response
message, just as any normal HTTP code in a Twisted Web server may do.
